### PR TITLE
fix(compiler): compile `.ngfactory.ts` files even if nobody reference…

### LINKED ISSUE #16741: NGC not type checking .ngfactory files on the first run
**Dan B in Ionic slack can give more info if needed**

[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

This exists in Angular 4.1.

If I have a type error in one of my templates, when I run `ngc` against it, it does not detect the error. After generating the `.ngfactory.ts` files, when I run `ngc` again, the error is properly detected.

For example, see building the `ionic-conference-app` demo with an error in a template. The two `ngc` calls are executed one-after-another.

```
dan at Dans-MacBook-Pro-2 in ~/Desktop/ionic-conference-app (master●)
$ ./node_modules/.bin/ngc

dan at Dans-MacBook-Pro-2 in ~/Desktop/ionic-conference-app (master●●)
$ ./node_modules/.bin/ngc
Error at /Users/dan/Desktop/ionic-conference-app/node_modules/ionic-angular/components/searchbar/searchbar.ngfactory.ts:237:33: Supplied parameters do not match any signature of call target.
Error at /Users/dan/Desktop/ionic-conference-app/node_modules/ionic-angular/components/searchbar/searchbar.ngfactory.ts:241:33: Supplied parameters do not match any signature of call target.
```

**Expected behavior**

To typecheck code the first time `ngc` is executed.

** Recreation steps **
1. Create a sample project
2. Introduce a typing error in the template
3. run `./node_modules/.bin/ngc` and observe the error in not reported.
4. run `./node_modules/.bin/ngc` again and observe the error is reported.



Thanks,
Dan

### PR DESCRIPTION
**Pull request by tbosch**
…s them.

This is especially important for library authors, as they will
not reference the .ngfactory.ts files.

Fixes #16741

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: packages/compiler-cli/test/main_spec.ts [hunk 1]
```diff
@@ -13,39 +13,53 @@ import * as path from 'path';
 
 import {main} from '../src/main';
 
+function getNgRootDir() {
+  const moduleFilename = module.filename.replace(/\\/g, '/');
+  const distIndex = moduleFilename.indexOf('/dist/all');
```
Comment (alexeagle) [inline packages/compiler-cli/test/main_spec.ts]: maybe throw if you don't find it? this failure will be hard to track down when we change the output layout next time

File: tools/@angular/tsc-wrapped/src/main.ts [hunk 2]
```diff
@@ -46,10 +51,18 @@ export function main(
     const basePath = path.resolve(process.cwd(), cliOptions.basePath || projectDir);
 
     // read the configuration options from wherever you store them
-    const {parsed, ngOptions} = tsc.readConfiguration(project, basePath, options);
+    let {parsed, ngOptions} = tsc.readConfiguration(project, basePath, options);
     ngOptions.basePath = basePath;
-    const createProgram = (host: ts.CompilerHost, oldProgram?: ts.Program) =>
-        ts.createProgram(parsed.fileNames, parsed.options, host, oldProgram);
+    let rootFileNames: string[] = parsed.fileNames.slice(0);
```
Comment (alexeagle) [inline tools/@angular/tsc-wrapped/src/main.ts]: I think we now prefer splat in a new array
`[...parsed.fileNames]`

File: tools/@angular/tsc-wrapped/src/main.ts [hunk 3]
```diff
@@ -46,10 +51,18 @@ export function main(
...
+    const createProgram = (host: ts.CompilerHost, oldProgram?: ts.Program) => {
+      return ts.createProgram(rootFileNames.slice(0), parsed.options, host, oldProgram);
```
Comment (alexeagle) [inline tools/@angular/tsc-wrapped/src/main.ts]: again

File: tools/@angular/tsc-wrapped/src/tsc.ts [hunk 4]
```diff
@@ -103,16 +103,19 @@ export function validateAngularCompilerOptions(options: AngularCompilerOptions):
...
 
 export class Tsc implements CompilerInterface {
-  public ngOptions: AngularCompilerOptions;
-  public parsed: ts.ParsedCommandLine;
-  private basePath: string;
+  private parseConfigHost: ts.ParseConfigHost;
 
-  constructor(private readFile = ts.sys.readFile, private readDirectory = ts.sys.readDirectory) {}
+  constructor(private readFile = ts.sys.readFile, private readDirectory = ts.sys.readDirectory) {
+    this.parseConfigHost = {
```
Comment (alexeagle) [inline tools/@angular/tsc-wrapped/src/tsc.ts]: can you give this a type now, to avoid typos? looks like we only omitted it due to g3 version skew earlier

**Review by alexeagle — APPROVED** [review]
thanks!

File: packages/compiler-cli/test/main_spec.ts (hunk 1 above)
Comment (tbosch) [inline packages/compiler-cli/test/main_spec.ts]: Done

File: tools/@angular/tsc-wrapped/src/main.ts (hunk 2 above)
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/main.ts]: Done

File: tools/@angular/tsc-wrapped/src/main.ts (hunk 3 above)
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/main.ts]: Done

File: tools/@angular/tsc-wrapped/src/tsc.ts (hunk 4 above)
Comment (tbosch) [inline tools/@angular/tsc-wrapped/src/tsc.ts]: It has one, see the member declaration.

**Review by mhevery — APPROVED** [review]

**Pull request merged**